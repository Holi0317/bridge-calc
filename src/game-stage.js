/**
 * Game stages
 * @enum
 */
export const gameStage = {
  /**
   * Waiting for bid result in this round.
   */
  waitingBid: 'waitingBid',
  /**
   * Waiting for win result in this round.
   */
  waitingWin: 'waitingWin',
  /**
   * Whole game has ended.
   */
  ended: 'ended'
}

export type GameStage = $Enum<gameStage>
