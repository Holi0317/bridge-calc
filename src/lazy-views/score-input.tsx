import * as React from "react";
import * as Loadable from "react-loadable";
import { connect } from "react-redux";
import { Loading } from "./loading";
import { IRootState } from "../types";
import { Redirect } from "react-router";

const importer = () =>
  import(
    "../score-input/layout" /* webpackChunkName: "score-input-view" */
  ).then(mod => mod.Layout);

export const Content = Loadable({
  loader: importer,
  loading: Loading
});

const mapStateToProps = (state: IRootState) => ({
  gameRedirect: state.currentGame == null
});

type LayoutProps = ReturnType<typeof mapStateToProps>;

export function LayoutImpl({ gameRedirect }: LayoutProps) {
  return gameRedirect ? <Redirect to="/entry" /> : <Content />;
}

export const ScoreInputView = connect(mapStateToProps)(LayoutImpl);
