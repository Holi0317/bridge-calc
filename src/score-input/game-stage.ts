/**
 * Enum of game stages
 */
export enum GameStage {
  /**
   * Waiting for bid result in this round.
   */
  waitingBid = "waitingBid",
  /**
   * Waiting for win result in this round.
   */
  waitingWin = "waitingWin",
  /**
   * Whole game has ended.
   */
  ended = "ended"
}
