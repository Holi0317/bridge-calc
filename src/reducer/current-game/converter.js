// @flow
import {GameStage} from '../../game-stage'
import {bidWinGenerator} from './bid-win-generator'
import type {WaitingBidState, WaitingWinState, EndedState} from './types'

export function toWaitingBidState(state: WaitingBidState | WaitingWinState): WaitingBidState {
  return {
    stage: GameStage.waitingBid,
    rounds: state.rounds,
    startTime: state.startTime,
    names: state.names,
    scores: state.scores,
    currentPlayerOrder: state.currentPlayerOrder,
    currentRound: state.currentRound,
    bid: state.bid
  }
}

export function toWaitingWinState(state: WaitingBidState | WaitingWinState): WaitingWinState {
  const newState = (toWaitingBidState(state): any) // Just to remove some whitelist boilerplate
  newState.stage = GameStage.waitingWin
  newState.win = (state: any).win || bidWinGenerator(Object.keys(state.names))
  return newState
}

export function toEndedState(state: WaitingBidState | WaitingWinState, endTime: Date): EndedState {
  return {
    stage: GameStage.ended,
    endTime,
    rounds: state.rounds,
    startTime: state.startTime,
    names: state.names,
    scores: state.scores
  }
}
