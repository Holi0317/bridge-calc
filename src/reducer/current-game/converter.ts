import {GameStage} from '../../game-stage'
import {bidWinGenerator} from './bid-win-generator'
import {IWaitingBidState, IWaitingWinState, IEndedState} from './types'

export function toWaitingBidState(state: IWaitingBidState | IWaitingWinState): IWaitingBidState {
  return {
    stage: GameStage.waitingBid,
    rounds: state.rounds,
    startTime: state.startTime,
    names: state.names,
    scores: state.scores,
    currentPlayerOrder: state.currentPlayerOrder,
    currentRound: state.currentRound,
    bid: state.bid
  }
}

export function toWaitingWinState(state: IWaitingBidState | IWaitingWinState): IWaitingWinState {
  const newState = toWaitingBidState(state) as any // Just to remove some whitelist boilerplate
  newState.stage = GameStage.waitingWin
  newState.win = (state as any).win || bidWinGenerator(Object.keys(state.names))
  return newState
}

export function toEndedState(state: IWaitingBidState | IWaitingWinState, endTime: Date): IEndedState {
  return {
    stage: GameStage.ended,
    endTime,
    rounds: state.rounds,
    startTime: state.startTime,
    names: state.names,
    scores: state.scores
  }
}
