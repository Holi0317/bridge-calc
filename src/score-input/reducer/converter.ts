import {GameStage} from '../game-stage'
import {bidWinGenerator} from './bid-win-generator'
import {IWaitingBidState, IWaitingWinState, IEndedState} from './types'

export function toWaitingBidState(state: IWaitingBidState | IWaitingWinState): IWaitingBidState {
  return {
    stage: GameStage.waitingBid,
    id: state.id,
    rounds: state.rounds,
    startTime: state.startTime,
    names: state.names,
    scores: state.scores,
    currentPlayerOrder: state.currentPlayerOrder,
    currentRound: state.currentRound,
    bid: state.bid
  }
}

export function toWaitingWinState(state: IWaitingBidState | IWaitingWinState): IWaitingWinState {
  const newState = toWaitingBidState(state) as any // Just to remove some whitelist boilerplate
  newState.stage = GameStage.waitingWin
  newState.win = (state as IWaitingWinState).win || bidWinGenerator(Object.keys(state.names))
  return newState
}

export function toEndedState(state: IWaitingBidState | IWaitingWinState, endTime: number): IEndedState {
  return {
    stage: GameStage.ended,
    endTime,
    id: state.id,
    rounds: state.rounds,
    startTime: state.startTime,
    names: state.names,
    scores: state.scores
  }
}
