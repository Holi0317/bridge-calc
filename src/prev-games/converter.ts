import {GameState} from '../score-input/reducer'
import {PrevGameEntry} from './types'

/**
 * Convert a game state to prev game object.
 * If the game state provided is null, an error will be thrown.
 * @exception {TypeError} - Given game state is null
 */
export function gameStateToPrevGame(gameState: GameState): PrevGameEntry {
  if (gameState == null) {
    throw new TypeError('[Prev games converter] game state must not be null for storage')
  }
  return gameState
}

/**
 * Convert a prev game object to game state.
 */
export function prevGameToGameState(prevGame: PrevGameEntry): GameState {
  return prevGame
}
