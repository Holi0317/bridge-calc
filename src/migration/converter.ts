import {IOldGameData, IOldPlayers, OldState} from './types'
import {PrevGameEntry} from '../prev-games/types'
import cuid from 'cuid'
import {IPlayerMap} from '../types'
import zipObject from 'lodash-es/zipObject'
import mapValues from 'lodash-es/mapValues'
import {GameStage} from '../score-input/game-stage'
import {toFront} from '../utils'
import {bidWinGenerator} from '../score-input/reducer/bid-win-generator'

interface IPlayerMaps {
  names: IPlayerMap<string>
  scores: IPlayerMap<number[]>
  bid: IPlayerMap<number | null>
  win: IPlayerMap<number | null>
}

/**
 * Migrate game state from v1.0.0 to new one.
 * This function may throw error (corrupted state?).
 * Use try-catch block when trying to convert.
 */
export function migrateOldState(oldState: IOldGameData): PrevGameEntry {
  if (oldState.state === OldState.notStarted) {
    throw new TypeError('[Migration] Old state is at notStarted state and cannot be migrated')
  }
  const playerProps = migratePlayers(oldState.players)
  const base = {
    id: cuid(),
    rounds: oldState.totalRounds,
    startTime: new Date().getTime(),
    names: playerProps.names,
    scores: playerProps.scores
  }

  const {state} = oldState
  const ids = Object.keys(playerProps.names)
  const currentPlayerOrder = toFront(ids, oldState.maker || 0)

  if (state === OldState.bid) {
    return {
      ...base,
      stage: GameStage.waitingBid,
      bid: playerProps.bid,
      currentPlayerOrder,
      currentRound: oldState.currentRound
    } as any // To suppress type error on some properties
  }

  if (state === OldState.inputWin) {
    return {
      ...base,
      stage: GameStage.waitingWin,
      bid: playerProps.bid,
      win: playerProps.win,
      currentPlayerOrder,
      currentRound: oldState.currentRound
    } as any // To suppress type error on some properties
  }

  if (state === OldState.waiting) {
    if (oldState.currentRound === oldState.totalRounds) {
      // Last round. Should return end game state
      return {
        ...base,
        stage: GameStage.ended,
        endTime: new Date().getTime()
      }
    }
    const maker = oldState.maker! + 1
    const order = toFront(ids, maker === ids.length ? 0 : maker)
    return {
      ...base,
      stage: GameStage.waitingBid,
      bid: bidWinGenerator(ids),
      currentRound: oldState.currentRound! + 1,
      currentPlayerOrder: order
    } as any
  }

  if (state === OldState.gameEnd) {
    return {
      ...base,
      stage: GameStage.ended,
      endTime: new Date().getTime()
    }
  }

  throw new Error('Convert of old game state failed. Unknown situation')
}

/**
 * Create names, scores, bid and win maps
 */
function migratePlayers(players: IOldPlayers[]): IPlayerMaps {
  const ids = players.map(cuid)
  const map = {
    names: 'name',
    scores: 'score',
    bid: 'bid',
    win: 'win'
  }
  const res = mapValues(map, (prop: keyof IOldPlayers) => (
    zipObject(ids, players.map(player => player[prop]))
  )) as any

  // Type convertion
  res.bid = numberize(res.bid)
  res.win = numberize(res.win)

  return res
}

export function numberize(obj: IPlayerMap<string | null>): IPlayerMap<number> {
  return mapValues(obj, b => b == null ? 0 : parseInt(b, 10))
}
