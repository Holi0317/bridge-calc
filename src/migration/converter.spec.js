import * as lolex from "lolex";
import { migrateOldState } from "./converter";
import { GameStage } from "../score-input/game-stage";

let clock = null;
beforeEach(() => {
  clock = lolex.install();
});
afterEach(() => {
  if (clock) {
    clock.uninstall();
  }
});

test("it should convert bid stage to waiting bid", () => {
  const data = {
    players: [
      {
        bid: null,
        win: null,
        _buffer: {
          lastRound: null,
          combo: null,
          miss: null,
          maxCombo: null,
          sum: null
        },
        score: [],
        name: "Rachel"
      },
      {
        bid: null,
        win: null,
        _buffer: {
          lastRound: null,
          combo: null,
          miss: null,
          maxCombo: null,
          sum: null
        },
        score: [],
        name: "Ruth"
      },
      {
        bid: null,
        win: null,
        _buffer: {
          lastRound: null,
          combo: null,
          miss: null,
          maxCombo: null,
          sum: null
        },
        score: [],
        name: "Helen"
      },
      {
        bid: null,
        win: null,
        _buffer: {
          lastRound: null,
          combo: null,
          miss: null,
          maxCombo: null,
          sum: null
        },
        score: [],
        name: "Walter"
      },
      {
        bid: null,
        win: null,
        _buffer: {
          lastRound: null,
          combo: null,
          miss: null,
          maxCombo: null,
          sum: null
        },
        score: [],
        name: "Patience"
      }
    ],
    maker: 0,
    currentRound: 1,
    state: 1,
    totalRounds: 6
  };
  const actual = migrateOldState(data);
  expect(actual.stage).toEqual(GameStage.waitingBid);
  expect(actual).toMatchSnapshot();
});

test("it should convert input win stage to waiting win", () => {
  const data = {
    players: [
      {
        bid: "1",
        win: null,
        _buffer: { lastRound: 0, combo: 0, miss: 0, maxCombo: 0, sum: 0 },
        score: [],
        name: "Frances"
      },
      {
        bid: "1",
        win: null,
        _buffer: { lastRound: 0, combo: 0, miss: 0, maxCombo: 0, sum: 0 },
        score: [],
        name: "John"
      },
      {
        bid: "0",
        win: null,
        _buffer: { lastRound: 0, combo: 0, miss: 0, maxCombo: 0, sum: 0 },
        score: [],
        name: "Patience"
      },
      {
        bid: "0",
        win: null,
        _buffer: { lastRound: 0, combo: 0, miss: 0, maxCombo: 0, sum: 0 },
        score: [],
        name: "Stephen"
      }
    ],
    maker: 0,
    currentRound: 1,
    state: 2,
    totalRounds: 13
  };
  const actual = migrateOldState(data);
  expect(actual.stage).toEqual(GameStage.waitingWin);
  expect(actual).toMatchSnapshot();
});

test("it should convert waiting stage to next round", () => {
  const data = {
    currentRound: 4,
    maker: 3,
    players: [
      {
        bid: "4",
        win: "3",
        _buffer: { lastRound: 3, combo: 0, miss: 2, maxCombo: 1, sum: 6 },
        score: [11, -1, -4, -1],
        name: "Christopher"
      },
      {
        bid: "2",
        win: "1",
        _buffer: { lastRound: 3, combo: 0, miss: 3, maxCombo: 0, sum: -6 },
        score: [-1, -1, -4, -1],
        name: "Rachel"
      },
      {
        bid: "0",
        win: "0",
        _buffer: { lastRound: 3, combo: 0, miss: 2, maxCombo: 1, sum: 5 },
        score: [10, -1, -4, 10],
        name: "Mark"
      },
      {
        bid: "0",
        win: "0",
        _buffer: { lastRound: 3, combo: 0, miss: 1, maxCombo: 2, sum: 11 },
        score: [10, 10, -9, 10],
        name: "Christopher"
      }
    ],
    state: 3,
    totalRounds: 13
  };
  const actual = migrateOldState(data);
  expect(actual.stage).toEqual(GameStage.waitingBid);
  expect(actual.currentRound).toEqual(5);
  expect(actual).toMatchSnapshot();
});

test("it should end game if waiting stage on last round", () => {
  const data = {
    players: [
      {
        bid: "2",
        win: "6",
        _buffer: { lastRound: 5, combo: 1, miss: 1, maxCombo: 3, sum: 62 },
        score: [11, 14, 19, -1, 19, -16],
        name: "Rachel"
      },
      {
        bid: "3",
        win: "0",
        _buffer: { lastRound: 5, combo: 2, miss: 3, maxCombo: 0, sum: 19 },
        score: [-1, -4, -4, 14, 14, -9],
        name: "Ruth"
      },
      {
        bid: "2",
        win: "0",
        _buffer: { lastRound: 5, combo: 0, miss: 2, maxCombo: 2, sum: 25 },
        score: [10, -4, 10, 10, -1, -4],
        name: "Helen"
      },
      {
        bid: "0",
        win: "0",
        _buffer: { lastRound: 5, combo: 3, miss: 1, maxCombo: 1, sum: 36 },
        score: [10, -4, 10, 10, 10, 10],
        name: "Walter"
      },
      {
        bid: "0",
        win: "0",
        _buffer: { lastRound: 5, combo: 3, miss: 1, maxCombo: 1, sum: 36 },
        score: [10, -4, 10, 10, 10, 10],
        name: "Patience"
      }
    ],
    maker: 0,
    currentRound: 6,
    state: 3,
    totalRounds: 6
  };
  const actual = migrateOldState(data);
  expect(actual.stage).toEqual(GameStage.ended);
  expect(actual).toMatchSnapshot();
});

test("it should convert round 5", () => {
  const data = {
    players: [
      {
        bid: "3",
        win: null,
        _buffer: { lastRound: 4, combo: 0, miss: 1, maxCombo: 3, sum: 43 },
        score: [11, 14, 19, -1],
        name: "Rachel"
      },
      {
        bid: "2",
        win: null,
        _buffer: { lastRound: 4, combo: 1, miss: 3, maxCombo: 0, sum: 5 },
        score: [-1, -4, -4, 14],
        name: "Ruth"
      },
      {
        bid: "1",
        win: null,
        _buffer: { lastRound: 4, combo: 2, miss: 1, maxCombo: 1, sum: 26 },
        score: [10, -4, 10, 10],
        name: "Helen"
      },
      {
        bid: "0",
        win: null,
        _buffer: { lastRound: 4, combo: 2, miss: 1, maxCombo: 1, sum: 26 },
        score: [10, -4, 10, 10],
        name: "Walter"
      },
      {
        bid: "0",
        win: null,
        _buffer: { lastRound: 4, combo: 2, miss: 1, maxCombo: 1, sum: 26 },
        score: [10, -4, 10, 10],
        name: "Patience"
      }
    ],
    maker: 4,
    currentRound: 5,
    state: 2,
    totalRounds: 6
  };
  const actual = migrateOldState(data);
  expect(actual.currentRound).toEqual(5);
  expect(actual.stage).toEqual(GameStage.waitingWin);
  expect(actual).toMatchSnapshot();
});

test("it should convert ended stage", () => {
  const data = {
    currentRound: null,
    maker: null,
    players: [
      {
        bid: null,
        win: null,
        _buffer: { lastRound: 2, combo: 2, miss: 0, maxCombo: 0, sum: 25 },
        score: [11, 14],
        name: "Christopher"
      },
      {
        bid: null,
        win: null,
        _buffer: { lastRound: 2, combo: 0, miss: 2, maxCombo: 0, sum: -2 },
        score: [-1, -1],
        name: "Rachel"
      },
      {
        bid: null,
        win: null,
        _buffer: { lastRound: 2, combo: 1, miss: 1, maxCombo: 0, sum: 9 },
        score: [-1, 10],
        name: "Mark"
      },
      {
        bid: null,
        win: null,
        _buffer: { lastRound: 2, combo: 1, miss: 1, maxCombo: 0, sum: 9 },
        score: [-1, 10],
        name: "Christopher"
      }
    ],
    state: 4,
    totalRounds: 2
  };
  const actual = migrateOldState(data);
  expect(actual.stage).toEqual(GameStage.ended);
  expect(actual).toMatchSnapshot();
});

test("it should throw error for notStarted state", () => {
  const data = {
    players: [],
    maker: null,
    currentRound: null,
    state: 0,
    totalRounds: 0
  };
  expect(() => {
    migrateOldState(data);
  }).toThrowError(TypeError);
});
