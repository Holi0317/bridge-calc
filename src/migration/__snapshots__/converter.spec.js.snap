// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should convert bid stage to waiting bid 1`] = `
Object {
  "bid": Object {
    "0": 0,
    "1": 0,
    "2": 0,
    "3": 0,
    "4": 0,
  },
  "currentPlayerOrder": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
  ],
  "currentRound": 1,
  "id": "5",
  "names": Object {
    "0": "Rachel",
    "1": "Ruth",
    "2": "Helen",
    "3": "Walter",
    "4": "Patience",
  },
  "rounds": 6,
  "scores": Object {
    "0": Array [],
    "1": Array [],
    "2": Array [],
    "3": Array [],
    "4": Array [],
  },
  "stage": "waitingBid",
  "startTime": 0,
}
`;

exports[`it should convert ended stage 1`] = `
Object {
  "endTime": 0,
  "id": "32",
  "names": Object {
    "28": "Christopher",
    "29": "Rachel",
    "30": "Mark",
    "31": "Christopher",
  },
  "rounds": 2,
  "scores": Object {
    "28": Array [
      11,
      14,
    ],
    "29": Array [
      -1,
      -1,
    ],
    "30": Array [
      -1,
      10,
    ],
    "31": Array [
      -1,
      10,
    ],
  },
  "stage": "ended",
  "startTime": 0,
}
`;

exports[`it should convert input win stage to waiting win 1`] = `
Object {
  "bid": Object {
    "6": 1,
    "7": 1,
    "8": 0,
    "9": 0,
  },
  "currentPlayerOrder": Array [
    "6",
    "7",
    "8",
    "9",
  ],
  "currentRound": 1,
  "id": "10",
  "names": Object {
    "6": "Frances",
    "7": "John",
    "8": "Patience",
    "9": "Stephen",
  },
  "rounds": 13,
  "scores": Object {
    "6": Array [],
    "7": Array [],
    "8": Array [],
    "9": Array [],
  },
  "stage": "waitingWin",
  "startTime": 0,
  "win": Object {
    "6": 0,
    "7": 0,
    "8": 0,
    "9": 0,
  },
}
`;

exports[`it should convert round 5 1`] = `
Object {
  "bid": Object {
    "22": 3,
    "23": 2,
    "24": 1,
    "25": 0,
    "26": 0,
  },
  "currentPlayerOrder": Array [
    "26",
    "22",
    "23",
    "24",
    "25",
  ],
  "currentRound": 5,
  "id": "27",
  "names": Object {
    "22": "Rachel",
    "23": "Ruth",
    "24": "Helen",
    "25": "Walter",
    "26": "Patience",
  },
  "rounds": 6,
  "scores": Object {
    "22": Array [
      11,
      14,
      19,
      -1,
    ],
    "23": Array [
      -1,
      -4,
      -4,
      14,
    ],
    "24": Array [
      10,
      -4,
      10,
      10,
    ],
    "25": Array [
      10,
      -4,
      10,
      10,
    ],
    "26": Array [
      10,
      -4,
      10,
      10,
    ],
  },
  "stage": "waitingWin",
  "startTime": 0,
  "win": Object {
    "22": 0,
    "23": 0,
    "24": 0,
    "25": 0,
    "26": 0,
  },
}
`;

exports[`it should convert waiting stage to next round 1`] = `
Object {
  "bid": Object {
    "11": 0,
    "12": 0,
    "13": 0,
    "14": 0,
  },
  "currentPlayerOrder": Array [
    "11",
    "12",
    "13",
    "14",
  ],
  "currentRound": 5,
  "id": "15",
  "names": Object {
    "11": "Christopher",
    "12": "Rachel",
    "13": "Mark",
    "14": "Christopher",
  },
  "rounds": 13,
  "scores": Object {
    "11": Array [
      11,
      -1,
      -4,
      -1,
    ],
    "12": Array [
      -1,
      -1,
      -4,
      -1,
    ],
    "13": Array [
      10,
      -1,
      -4,
      10,
    ],
    "14": Array [
      10,
      10,
      -9,
      10,
    ],
  },
  "stage": "waitingBid",
  "startTime": 0,
}
`;

exports[`it should end game if waiting stage on last round 1`] = `
Object {
  "endTime": 0,
  "id": "21",
  "names": Object {
    "16": "Rachel",
    "17": "Ruth",
    "18": "Helen",
    "19": "Walter",
    "20": "Patience",
  },
  "rounds": 6,
  "scores": Object {
    "16": Array [
      11,
      14,
      19,
      -1,
      19,
      -16,
    ],
    "17": Array [
      -1,
      -4,
      -4,
      14,
      14,
      -9,
    ],
    "18": Array [
      10,
      -4,
      10,
      10,
      -1,
      -4,
    ],
    "19": Array [
      10,
      -4,
      10,
      10,
      10,
      10,
    ],
    "20": Array [
      10,
      -4,
      10,
      10,
      10,
      10,
    ],
  },
  "stage": "ended",
  "startTime": 0,
}
`;
